components/rollup_node/block_gen.md

```diff
@@ -0,0 +1,147 @@
+    l1TxOrigin:  Address
+    target:      Address
+    data:        bytes
+
+class ContextDeposit(Deposit):
+    blockHash:   bytes32
+    blockNumber: uint64
+    timestamp:   uint64
+    baseFee:     uint64
+```
+
+### Deposit Block Generation
+
+The function `generate_deposit_block(l1_block_number)` is defined as:
+
+- Generate the `ContextDeposit` using the L1 block body. Specifically the L1 block's `blockHash`, `blockNumber`, `timestamp`, and `baseFee`.
+- Get all events emitted by the `DepositFeed` contract at block `l1_block_number`.
+    - For each event, generate a new `UserDeposit` based on the emitted calldata. All fields should be emitted by DepositFeed contract.
+
+#### **Footnote**: The DepositFeed ensures that no matter the size of the event it is possible to prove the deposit data to the fraud proof. See the DepositFeed spec for details.
+
+## Sequencer Blocks in Depth
+
+Sequencer blocks contain a majority of the transactions submitted to the rollup chain. Sequencer blocks can also influence the current timestamp of L2 which is exposed to contracts & is used in the EIP1559 fee pricing calculations.
+
+Sequencer blocks are of the following form:
+
+```python
+class SequencerBlock:
+    target_epoch: uint64  # epoch number that this block is intended for
+    sequencer_suggested_timestamp: uint64  # a timestamp suggested by the sequencer
+    transactions: List[Transaction]  # A list of transactions in the block
+```
+
+These sequencer blocks must be validated as properly formatted. These validity checks are:
+
+1. `target_epoch` must be within the sequencing window (introduced in the previous sections)
+2. `sequencer_suggested_timestamp` must be monotonic & within an acceptable range.
+3. `transactions` must all be correctly formatted (eg. contain valid signatures).
+
+If any of these checks fail the block must be skipped or ignored. Otherwise, all transactions in the sequencer block will be executed.
+
+## Timestamps in Depth
+
+Time can be manipulated by the sequencer; however, it must stay within reasonable bounds and must be monotonic (ie. can never go backwards). The specific conditions that must be met are:
+
+```python
+# last_l1_timestamp is the L1 timestamp set during the last ContextDeposit
+# last_timestamp is the timestamp from the previous rollup block
+# average_l1_block_time is the average time between two L1 blocks
+# sequencing_window_size is the number of L1 blocks the sequencer has to submit their blocks
+min_timestamp = max(last_l1_timestamp, last_timestamp)
+max_timestamp = last_l1_timestamp + sequencing_window_size * average_l1_block_time
+
+assert sequencer_suggested_timestamp >= min_timestamp
+assert sequencer_suggested_timestamp <= max_timestamp
+```
+
+The timestamp exposed on the rollup chain is:
+
+```python
+# last_sequencer_suggested_timestamp is the most recent valid sequencer_suggested_timestamp
+def timestamp():
+    max(last_l1_timestamp, last_sequencer_suggested_timestamp)
```

Comment:
I think that setting the value in this way contradicts above where we say it's a validity condition/assertion. FWIW I'm more of a fan of setting the timestamp for the sequencer if they screw up, seems like a better UX to still apply the transactions albeit "late."

Reply:
Ah so I intended for this function to be used to ensure the `timestamp` increases even when there haven't been sequencer feed elements in a very long time. So the validity condition only relates to sequencer feed elements, but this function is used to determine the block timestamp at all times.

Reply:
Ah, got it. I would be in favor of making the `timestamp()` function purely on the *current* epoch sequencer suggestion (if present) and the last l2 timestamp. If I'm understanding your proposal right, then `last_sequencer_suggested_timestamp` would otherwise have to be a new thing persisted in state, correct?

Reply:
Hmm so I'm having trouble figuring out how to get around persisting `last_sequencer_suggested_timestamp` in state. I think you need either `last_sequencer_suggested_timestamp` or you need to persist `last_timestamp`. This covers the case where:

1. Deposit block comes in setting `last_l1_timestamp=1`
1. Sequencer sets timestamp to `last_l1_timestamp + 10`
1. Sequencer goes offline
1. Deposit block is processed which updates increases the `last_l1_timestamp` to `5`

What is `timestamp` in this case? Presumably the timestamp would not increase past the sequencer suggested timestamp until the `last_l1_timestamp` increases naturally to greater than `10`.

I figured it would be a little cleaner to store the `last_sequencer_timestamp` vs storing `last_l2_timestamp` because it gives you a better sense of where the timestamp number is coming from. However, I'm not super opinionated.

Reply:
Seems `last_l2_timestamp` is better because we don't have to store an extra value and can just look back into the previous block. That said, it's an implementation detail

Reply:
Maybe `last_l2_timestamp + 1` so that two blocks never have the same timestamp